hbase-client/src/main/java/org/apache/hadoop/hbase/client/MetaReplicaLoadBalanceReplicaSimpleChooser.java

```diff
@@ -0,0 +1,248 @@
+import java.util.concurrent.ConcurrentSkipListMap;
+import java.util.concurrent.ThreadLocalRandom;
+import static org.apache.hadoop.hbase.HConstants.DEFAULT_META_REPLICA_NUM;
+import static org.apache.hadoop.hbase.HConstants.META_REPLICAS_NUM;
+import static org.apache.hadoop.hbase.client.ConnectionUtils.isEmptyStopRow;
+import static org.apache.hadoop.hbase.util.Bytes.BYTES_COMPARATOR;
+import static org.apache.hadoop.hbase.util.ConcurrentMapUtils.computeIfAbsent;
+
+/**
+ * MetaReplicaLoadBalanceReplicaSimpleChooser implements a simple meta replica load balancing
+ * algorithm. It maintains a stale location cache for each table. Whenever client looks up meta,
+ * it first check if the row is the stale location cache, if yes, this means the the location from
+ * meta replica is stale, it will go to the primary meta to look up update-to-date location;
+ * otherwise, it will randomly pick up a meta replica region for meta lookup. When clients receive
+ * RegionNotServedException from region servers, it will add these region locations to the stale
+ * location cache. The stale cache will be cleaned up periodically by a chore.
+ */
+
+/**
+ * StaleLocationCacheEntry is the entry when a stale location is reported by an client.
+ */
+class StaleLocationCacheEntry {
+  // meta replica id where
+  private int metaReplicaId;
+
+  // timestamp in milliseconds
+  private long timestamp;
+
+  private byte[] endKey;
+
+  StaleLocationCacheEntry(final int metaReplicaId, final byte[] endKey) {
+    this.metaReplicaId = metaReplicaId;
+    this.endKey = endKey;
+    timestamp = System.currentTimeMillis();
+  }
+
+  public byte[] getEndKey() {
+    return this.endKey;
+  }
+
+  public int getMetaReplicaId() {
+    return this.metaReplicaId;
+  }
+  public long getTimestamp() {
+    return this.timestamp;
+  }
+
+  @Override
+  public String toString() {
+    return new ToStringBuilder(this)
+      .append("endKey", endKey)
+      .append("metaReplicaId", metaReplicaId)
+      .append("timestamp", timestamp)
+      .toString();
+  }
+}
+
+/**
+ * A simple implementation of MetaReplicaLoadBalanceReplicaChooser.
+ *
+ * It follows a simple algorithm to choose a meta replica to go:
+ *
+ *  1. If there is no stale location entry for rows it looks up, it will randomly
+ *     pick a meta replica region to do lookup.
+ *  2. If the location from meta replica region is stale, client gets RegionNotServedException
+ *     from region server, in this case, it will create StaleLocationCacheEntry in
+ *     MetaReplicaLoadBalanceReplicaSimpleChooser.
+ *  3. When client tries to do meta lookup, it checks StaleLocationCache first for rows it tries to
+ *     lookup, if entry exists, it will go with primary meta region to do lookup; otherwise, it
+ *     will follow step 1.
+ *  4. A chore will periodically run to clean up cache entries in the StaleLocationCache.
+ */
+class MetaReplicaLoadBalanceReplicaSimpleChooser implements MetaReplicaLoadBalanceReplicaChooser {
```

Comment:
Maybe first we just need to cache the replica id which contains the stale data? For example, we get location from replica 1 and it is stale, but it does not mean that the location from replica 2 will be stale? If we use the algorithm here, we will go to replica 0(the primary replica) instead of replica 1.

And consider the whole client locator architecture, where have a meta cache in the upper layer, I would suggest that, we only provide two selection ways here: the first is random(no primay), and if you got a stale data, then use the second way, only go to primary. Maybe we could add a parameter on the choose method to indicate whether we only want to go to the primary replica.

Reply:
Our original plan is to come up with a simple algorithm first.
1. Select a meta replica region first for lookup, if the data is stale, then go to primary.

Later, maybe we can experiment some smarter way about how to select the next meta replica in case of error.
Here,  state needed to make decision is contained insider this selector class. If I understand correctly, to have a flag in choose method indicates that top layer needs to keep some state for this purpose. Trying to avoid that so there is no side effect if this mode is not enabled.

Reply:
For a normal case, > 99% of region locations from meta replica will be up to date. In extreme cases such as region server crash, it will depends on how fast replication catches up. Will try to add more tests to find out.

Reply:
Will do.